File: conformance/work-class/1.0.0-draft.2/typescript/adapter.ts
import { pathToFileURL } from 'node:url';
import { AdmissionError, canonical, decodeArtifact, parseRestricted } from './canonical.ts';
import { evaluateAggregate, validateDefinition } from './aggregate.ts';
import { evaluateAuthority, validateAuthorityDefinition } from './authority.ts';
import { checkEvidence, readback, validateBoundary, validateEvidenceArtifact } from './evidence.ts';
import { deploy, step } from './lifecycle.ts';
import { aggregateStep } from './reservations.ts';
import { LIMITS, PROTOCOL, requireShape, selectedCandidate, VERSION } from './schema.ts';
import { validateWorkClass } from './work-class.ts';
import type { RecordValue as Row } from './values.ts';

const OPERATIONS = ['aggregate-step', 'capabilities', 'check-evidence', 'deploy', 'evaluate', 'readback', 'step', 'validate'];
const ROLES = ['AGGREGATE_EVALUATOR', 'AUTHORITY_EVALUATOR', 'CHOICE_LOOP_RUNTIME', 'DEADLINE_RUNTIME', 'LINEAR_WORK_RUNTIME', 'PARALLEL_FANOUT_RUNTIME', 'REFERENCE_SINGLE_PROCESS_CONTENTION', 'REVIEW_EVIDENCE', 'SHARED_BUDGET_TRANSITIONS'];
const CLAIMED_ROLES = process.env.WORK_CLASS_ADAPTER_ROLES === undefined
  ? ROLES
  : process.env.WORK_CLASS_ADAPTER_ROLES.split(',').filter(Boolean).sort();
const CLAIMED_OPERATIONS = CLAIMED_ROLES.includes('SHARED_BUDGET_TRANSITIONS') ? OPERATIONS : OPERATIONS.filter(operation => operation !== 'aggregate-step');
function object(value: unknown): value is Row { return value !== null && typeof value === 'object' && !Array.isArray(value); }

function operationResult(operation: string, input: Row, pin: string): unknown {
  if (operation === 'capabilities') {
    requireShape('protocol.schema.json', 'capabilitiesInput', input);
    return { status: 'CAPABILITIES', protocol: PROTOCOL, specification_pin: pin, roles: CLAIMED_ROLES, operations: CLAIMED_OPERATIONS, tzdb: '2025a', limits: LIMITS };
  }
  if (operation === 'deploy') return deploy(input, pin);
  if (operation === 'step') return step(input, pin);
  if (operation === 'check-evidence') return checkEvidence(input, pin);
  if (operation === 'aggregate-step') return CLAIMED_ROLES.includes('SHARED_BUDGET_TRANSITIONS') ? aggregateStep(input, pin) : { status: 'REFUSED', code: 'ROLE_UNSUPPORTED', path: '' };
  if (operation === 'evaluate') {
    requireShape('protocol.schema.json', 'evaluateInput', input);
    return input.kind === 'AUTHORITY' ? evaluateAuthority(input.request, pin) : evaluateAggregate(input.request, pin);
  }
  requireShape('protocol.schema.json', 'artifactInput', input);
  let artifact: unknown;
  try { artifact = decodeArtifact(input.artifact_bytes_base64).value; }
  catch (error) {
    if (!(error instanceof AdmissionError)) throw error;
    throw new AdmissionError(error.code === 'NONCANONICAL' || error.code === 'NUMBER_FORBIDDEN' ? 'ARTIFACT_NONCANONICAL' : 'ARTIFACT_ENCODING_INVALID');
  }
  if (input.specification_pin !== pin) throw new AdmissionError('VERSION_UNSUPPORTED');
  const suffix: Record<string, string> = {
    AGGREGATE_DEFINITION: '/aggregate-definition', AUTHORITY_DEFINITION: '/authority-definition', WORK_CLASS: '/work-class-definition',
    BOUNDARY_RECORD: '/boundary-record', EVIDENCE_RECORD: '/evidence-record',
  };
  if (object(artifact) && typeof artifact.schema === 'string' && artifact.schema !== VERSION + suffix[input.kind]) throw new AdmissionError('VERSION_UNSUPPORTED');
  let identity: string;
  if (input.kind === 'AGGREGATE_DEFINITION') identity = validateDefinition(artifact).identity;
  else if (input.kind === 'AUTHORITY_DEFINITION') {
    const result = validateAuthorityDefinition(artifact);
    if (result.status !== 'ACCEPTED') throw new AdmissionError(result.code);
    identity = result.digest;
  } else if (input.kind === 'WORK_CLASS') identity = validateWorkClass(artifact, pin).digest;
  else if (input.kind === 'BOUNDARY_RECORD') identity = validateBoundary(artifact, pin);
  else identity = validateEvidenceArtifact(artifact);
  return operation === 'validate' ? { status: 'ACCEPTED', digest: identity } : { status: 'READBACK', digest: identity, lines: readback(artifact) };
}

export function respond(value: unknown, pin: string): Row {
  const requestId = object(value) && typeof value.request_id === 'string' ? value.request_id : null;
  const operation = object(value) && typeof value.operation === 'string' ? value.operation : null;
  const responseOperation = operation;
  if (!object(value) || !operation || !OPERATIONS.includes(operation) || !object(value.input) || value.protocol !== PROTOCOL || !('request_id' in value) || Object.keys(value).sort().join('|') !== 'input|operation|protocol|request_id') {
    return { protocol: PROTOCOL, request_id: requestId, operation: responseOperation, result: { status: 'REFUSED', code: 'PROTOCOL_INVALID', path: '' } };
  }
  try {
    return { protocol: PROTOCOL, request_id: requestId, operation, result: operationResult(operation, value.input, pin) };
  } catch (error) {
    if (!(error instanceof AdmissionError)) throw error;
    return { protocol: PROTOCOL, request_id: requestId, operation, result: { status: 'REFUSED', code: error.code === 'TIME_INVALID' ? 'INPUT_INVALID' : error.code, path: '' } };
  }
}

async function main(): Promise<void> {
  const pin = selectedCandidate();
  let pending = Buffer.alloc(0);
  for await (const chunk of process.stdin) {
    pending = Buffer.concat([pending, chunk]);
    let boundary: number;
    while ((boundary = pending.indexOf(10)) >= 0) {
      const line = pending.subarray(0, boundary); pending = pending.subarray(boundary + 1);
      const decoded = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(line);
      parseRestricted(decoded);
      process.stdout.write(canonical(respond(JSON.parse(decoded), pin)) + '\n');
    }
  }
  if (pending.length) throw new Error('Transport request lacks its final line-feed');
}
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) main().catch(error => { process.stderr.write(String(error) + '\n'); process.exitCode = 1; });

File: conformance/work-class/1.0.0-draft.2/python/adapter.py
#!/usr/bin/env python3
"""Restricted JSON-lines adapter for the independent draft-2 consumer."""

from __future__ import annotations

import os
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parent
for dependency in (ROOT, ROOT / "vendor"):
    if str(dependency) not in sys.path:
        sys.path.insert(0, str(dependency))

from implementation import artifacts, evidence, operations  # noqa: E402
from implementation.common import (  # noqa: E402
    IDENTIFIER_PATTERN,
    PIN,
    PROTOCOL,
    Refusal,
    artifact_bytes,
    canonical,
    parse_json,
    schema_valid,
)


ALL_OPERATIONS = (
    "aggregate-step",
    "capabilities",
    "check-evidence",
    "deploy",
    "evaluate",
    "readback",
    "step",
    "validate",
)
ALL_ROLES = (
    "AGGREGATE_EVALUATOR",
    "AUTHORITY_EVALUATOR",
    "CHOICE_LOOP_RUNTIME",
    "DEADLINE_RUNTIME",
    "LINEAR_WORK_RUNTIME",
    "PARALLEL_FANOUT_RUNTIME",
    "REFERENCE_SINGLE_PROCESS_CONTENTION",
    "REVIEW_EVIDENCE",
    "SHARED_BUDGET_TRANSITIONS",
)
ACTIVE_ROLES = tuple(sorted(filter(None,os.environ.get("WORK_CLASS_ADAPTER_ROLES",",".join(ALL_ROLES)).split(",")),key=str.encode))
IMPLEMENTED_OPERATIONS = ALL_OPERATIONS if "SHARED_BUDGET_TRANSITIONS" in ACTIVE_ROLES else tuple(operation for operation in ALL_OPERATIONS if operation!="aggregate-step")
LIMITS = {
    "max_definitions": "64",
    "max_operations": "256",
    "max_collection_members": "1024",
    "max_numeric_digits": "128",
    "max_expression_depth": "64",
    "max_expression_nodes": "4096",
}


def refusal(code: str):
    return {"status": "REFUSED", "code": code, "path": ""}


def _echo_identifier(value):
    return value if isinstance(value, str) and IDENTIFIER_PATTERN.fullmatch(value) else None


def _response(request_id, operation, result):
    return {
        "protocol": PROTOCOL,
        "request_id": request_id,
        "operation": operation,
        "result": result,
    }


def _protocol_error(request):
    if not isinstance(request, dict):
        return _response(None, None, refusal("PROTOCOL_INVALID"))
    return _response(
        _echo_identifier(request.get("request_id")),
        _echo_identifier(request.get("operation")),
        refusal("PROTOCOL_INVALID"),
    )


def _dispatch(operation: str, value):
    if operation == "capabilities":
        if value != {}:
            return refusal("SCHEMA_INVALID")
        return {
            "status": "CAPABILITIES",
            "protocol": PROTOCOL,
            "specification_pin": PIN,
            "roles": list(ACTIVE_ROLES),
            "operations": list(IMPLEMENTED_OPERATIONS),
            "tzdb": "2025a",
            "limits": LIMITS,
        }
    if operation in ("validate", "readback"):
        if not schema_valid(value, "protocol.schema.json", "artifactInput"):
            return refusal("SCHEMA_INVALID")
        if value["specification_pin"] != PIN:
            return refusal("VERSION_UNSUPPORTED")
        try:
            artifact = artifact_bytes(value["artifact_bytes_base64"])
            if operation == "validate":
                return {"status": "ACCEPTED", "digest": artifacts.admit(value["kind"], artifact)}
            return artifacts.readback(value["kind"], artifact)
        except Refusal as error:
            return refusal(error.code)
    if operation == "evaluate":
        if not schema_valid(value, "protocol.schema.json", "evaluateInput"):
            return refusal("SCHEMA_INVALID")
        if value["kind"] == "AGGREGATE":
            return operations.aggregate(value["request"])
        return operations.authority_evaluate(value["request"])
    if operation == "aggregate-step":
        return operations.reservation_step(value) if "SHARED_BUDGET_TRANSITIONS" in ACTIVE_ROLES else refusal("ROLE_UNSUPPORTED")
    if operation == "check-evidence":
        return evidence.check(value)
    if operation == "deploy":
        return operations.lifecycle_deploy(value)
    if operation == "step":
        return operations.lifecycle_step(value)
    raise RuntimeError("unreachable operation")


def handle(request):
    if (
        type(request) is not dict
        or set(request) != {"protocol", "request_id", "operation", "input"}
        or request.get("protocol") != PROTOCOL
        or _echo_identifier(request.get("request_id")) is None
        or _echo_identifier(request.get("operation")) is None
        or request["operation"] not in ALL_OPERATIONS
        or type(request.get("input")) is not dict
    ):
        return _protocol_error(request)
    result = _dispatch(request["operation"], request["input"])
    response = _response(request["request_id"], request["operation"], result)
    if not schema_valid(response, "protocol.schema.json", "response"):
        raise RuntimeError("operation produced an invalid protocol response")
    return response


def main():
    if sys.argv[1:]:
        raise SystemExit("usage: adapter.py")
    for raw in sys.stdin.buffer:
        if not raw.endswith(b"\n"):
            print("unterminated request", file=sys.stderr)
            raise SystemExit(2)
        try:
            request = parse_json(raw[:-1])
            response = handle(request)
        except Exception as error:
            print(f"transport or execution failure: {error}", file=sys.stderr)
            raise SystemExit(2) from error
        sys.stdout.buffer.write(canonical(response) + b"\n")
        sys.stdout.buffer.flush()


if __name__ == "__main__":
    main()
